cmd/aries-agent-mobile/pkg/wrappers/rest/aries.go

```diff
@@ -42,7 +54,121 @@ func NewAries(opts *config.Options) (*Aries, error) {
 
 	endpoints := getControllerEndpoints()
 
-	return &Aries{endpoints: endpoints, URL: opts.AgentURL, Token: opts.APIToken}, nil
+	a := &Aries{
+		endpoints:    endpoints,
+		URL:          opts.AgentURL,
+		Token:        opts.APIToken,
+		WebsocketURL: opts.WebsocketURL,
+		subscribers:  make(map[string]map[string][]api.Handler),
+	}
+
+	go a.startNotificationListener()
+
+	return a, nil
+}
+
+// Incoming represents WebSocket event message.
+type Incoming struct {
+	ID      string      `json:"id"`
+	Topic   string      `json:"topic"`
+	Message interface{} `json:"message"`
+}
+
+func (ar *Aries) startNotificationListener() { // nolint: gocyclo,funlen
+	if ar.WebsocketURL == "" {
+		return
+	}
+
+	conn, _, err := websocket.Dial(context.Background(), ar.WebsocketURL, &websocket.DialOptions{ //nolint:bodyclose
+		HTTPClient: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true, // nolint: gosec
+				},
+			},
+		},
+	})
+	if err != nil {
+		logger.Errorf("notification listener: websocket dial: %v", err)
+
+		return
+	}
+
+	defer func() {
+		err = conn.Close(websocket.StatusNormalClosure, "closing the connection")
+		if err != nil && websocket.CloseStatus(err) != websocket.StatusNormalClosure {
+			logger.Errorf("notification listener: close connection: %v", err)
+		}
+	}()
+
+	for {
+		var incoming *Incoming
+		// listens for notifications
+		if err := wsjson.Read(context.Background(), conn, &incoming); err != nil {
+			if websocket.CloseStatus(err) != websocket.StatusNormalClosure {
+				logger.Errorf("notification listener: read: close connection: %v", err)
+			}
+
+			// exit if websocket is closed
+			if websocket.CloseStatus(err) != -1 {
+				return
+			}
+
+			continue
+		}
+
+		ar.mutex.RLock()
```

Comment:
should defer `RUnlock` here

Reply:
it is a loop. `defer ` will not work

Reply:
`defer ar.mutex.RUnlock()` will not work because what follows is a loop?

Reply:
it fixed